README
= attachment-fu

attachment_fu is a plugin by Rick Olson (aka technoweenie
http://techno-weenie.net) and is the successor to acts_as_attachment. To get a
basic run-through of its capabilities, check out {Mike Clark's
tutorial}[http://clarkware.com/cgi/blosxom/2007/02/24#FileUploadFu].

= attachment_fu functionality

attachment_fu facilitates file uploads in Ruby on Rails. There are a few
storage options for the actual file data, but the plugin always at a minimum
stores metadata for each file in the database.

There are four storage options for files uploaded through attachment_fu:

* File system
* Database file
* Amazon S3
* Rackspace (Mosso) Cloud Files

Each method of storage many options associated with it that will be covered in
the following section. Something to note, however, is that the Amazon S3 storage
requires you to modify +config/amazon_s3.yml+, the Rackspace Cloud Files storage
requires you to modify +config/rackspace_cloudfiles.yml+, and the Database file
storage requires an extra table.

= attachment_fu models

For all three of these storage options a table of metadata is required. This
table will contain information about the file (hence the 'meta') and its
location. This table has no restrictions on naming, unlike the extra table
required for database storage, which must have a table name of +db_files+ (and
by convention a model of +DbFile+).

Two methods are available to models: +has_attachment+ and
+validates_as_attachment+.

== has_attachment(options = {})

This method accepts the options in a hash:

[:content_type]
    Allowed content types.

    By default, all content types are allowed. Use +:image+ to allow all
    standard image types.
    

[:min_size]
    Minimum file size.
    By default, set to +1.byte+.

[:max_size]
    Maximum file size.

    By default, set to +1.megabyte+.

[:size]
    Minimum and maximum file size.

    By default, set to +1..1.megabyte+. Overrides +:min_size+ and
    +:max_size+.

[:resize_to]
    Used by RMagick.

    Tells RMagick how to resize images. Pass either an array specifying
    width and height or a geometry string.  Prefixing the geometry string
    with a 'c' will crop the image to the specified size.

[:sharpen_on_resize]
    Used by RMagick.

    If set to true, images are sharpened after being resized.

[:thumbnails]
    A set of thumbnails to generate.

    This accepts a hash of filename suffixes and RMagick resizing options. This
    option need only be included if you want thumbnailing.

    If you have a polymorphic parent relationship, you can provide
    parent-type-specific thumbnail settings by using a pair with the type string
    as key and a Hash of thumbnail definitions, or a method symbol, as value.
    The method symbol will call the named method in order to get a
    dynamically-built Hash of thumbnail definitions, which gives you full
    flexibility. AttachmentFu automatically detects your first polymorphic
    +belongs_to+ relationship.

[:thumbnail_class]
    Which model class to use for thumbnails.

    By default, the current attachment class is used.

[:jpeg_quality]
    JPEG quality settings for thumbnail resizes.

    Arguments can be in multiple formats:

    * Integer from 0 (basically crap) to 100 (basically lossless, fat files).

    * When relying on tdd-image_science, you can also use one of its +JPEG_xxx+
      constants for predefined ratios/settings.

    * You can also use a Hash, with keys being either  thumbnail symbols (I
      repeat: _symbols_) or surface boundaries.  A surface boundary is a string
      starting with either '<' or '>=', followed by a number of pixels.  This
      lets you specify per-thumbnail or per-general-thumbnail-"size" JPEG
      qualities. (which can be useful when you have a _lot_ of thumbnail
      options). Surface example: <code>{ '<2000' => 90, '>=2000' => 75 }</code>.

    Defaults vary depending on the processor (ImageScience: 100%,
    Rmagick/MiniMagick/Gd2: 75%, CoreImage: auto-adjust).  Note that only
    tdd-image_science (available from GitHub) currently supports explicit JPEG
    quality; the default image_science currently forces 100%.

[:path_prefix]
    Path to store the uploaded files in. Uses <code>public/#{table_name}</code>
    by default for the filesystem, and just #{table_name} for the S3 and Cloud
    Files backend. Setting this sets the +:storage+ to +:file_system+.

[:partition]
    Whether to partiton files in directories like +/0000/0001/image.jpg+.
    Default is true. Only applicable to the +:file_system+ backend.

[:storage]
    Specifies the storage system to use. Defaults to +:db_file+. Options are
    +:file_system+, +:db_file+, +:s3+, and +:cloud_files+.

[:cloudfront]
    If using S3 for storage, this option allows for serving the files via Amazon
    CloudFront. Defaults to false.

[:processor]
    Sets the image processor to use for resizing of the attached image. Options
    include ImageScience, Rmagick, MiniMagick, Gd2 and CoreImage. Default is
    whatever is installed.

[:uuid_primary_key]
    If your model's primary key is a 128-bit UUID in hexadecimal format, then
    set this to true.

[:association_options]
    attachment_fu automatically defines associations with thumbnails with
    +has_many+ and +belongs_to+. If there are any additional options that you
    want to pass to these methods, then specify them here.

Examples:

    has_attachment max_size: 1.kilobyte
    has_attachment size: 1.megabyte..2.megabytes
    has_attachment content_type: 'application/pdf'
    has_attachment content_type: ['application/pdf', 'application/msword', 'text/plain']
    has_attachment content_type: :image, resize_to: [50,50]
    has_attachment content_type: ['application/pdf', :image], resize_to: 'x50'
    has_attachment thumbnails: { thumb: [50, 50], geometry: 'x50' }
    has_attachment storage: :file_system, path_prefix: 'public/files'
    has_attachment(
        storage: :file_system,
        path_prefix: 'public/files',
        content_type: :image,
        resize_to: [50, 50],
        partition: false
    )
    has_attachment(
        storage: :file_system,
        path_prefix: 'public/files',
        thumbnails: { thumb: [50, 50], geometry: 'x50' }
    )
    has_attachment storage: :s3
    has_attachment store: :s3, cloudfront: true
    has_attachment storage: :cloud_files

    # Let's say we have a polymorphic belongs_to, e.g. called 'imageable', where
    # imageable_type (or whatever the :foreign_type option was set to) can be,
    # among other things, 'Product', 'User' or 'Editorial', each of which should
    # have extra thumbnails:

    has_attachment thumbnails: {
        thumb: [50, 50],
        geometry: 'x50',
        products: { large_thumb: '169x169!', zoomed: '500x500>' },
        editorials: { fullsize: '150x100>' },
        users: { avatar: '64x64!' }
    }

    # JPEG qualities…

    has_attachment jpeg_quality: 75
    has_attachment jpeg_quality: 80 | ImageScience::JPEG_PROGRESSIVE
    has_attachment(
        thumbnails: { thumb: [50, 50], geometry: 'x50' },
        jpeg_quality: { nil => 75, thumb: 90, geometry: 90 }
    )
    has_attachment(
        thumbnails: { thumb: [50, 50], geometry: 'x50' },
        jpeg_quality: { '<2000' => 90, '>=2000' => 75 }
    )

== validates_as_attachment

This method prevents files outside of the valid range (+:min_size+ to
+:max_size+, or the +:size+ range) from being saved. It does not however, halt
the upload of such files. They will be uploaded into memory regardless of size
before validation.

To perform this validation, simply add +validates_as_attachment+ to your model.

= attachment_fu migrations

Fields for attachment_fu metadata tables…

In general:

    size,         :integer # file size in bytes
    content_type, :string  # mime type, ex: application/mp3
    filename,     :string  # sanitized filename

That reference images:

    height,       :integer # in pixels
    width,        :integer # in pixels

That reference images that will be thumbnailed:

    parent_id,    :integer  # id of parent image (on the same table, a
                            # self-referencing foreign-key). Only populated if
                            # the current object is a thumbnail.
    thumbnail,    :string   # The type of thumbnail this attachment record
                            # describes. Only populated if the current object is
                            # a thumbnail. Example:
                            #
                            # (In Model 'Avatar')
                            # has_attachment(
                            #     :content_type => :image,
                            #     :storage => :file_system,
                            #     :max_size => 500.kilobytes,
                            #     :resize_to => '320x200>',
                            #     :thumbnails => {
                            #         :small => '10x10>',
                            #         :thumb => '100x100>'
                            #     }
                            # )
                            #
                            # (Elsewhere)
                            # @user.avatar.thumbnails.first.thumbnail # => 'small'
                            #
    db_file_id,   :integer  # ID of the file in the database (foreign key) that
                            # reference files stored in the database (:db_file).

Field for attachment_fu +db_files+ table:

    data, :binary # binary file data, for use in database file storage

= attachment_fu views

There are two main views tasks that will be directly affected by attachment_fu:
upload forms and displaying uploaded images.

There are two parts of the upload form that differ from typical usage.

1. Include 'multipart: true' in the html options of the form_for tag. Example:

        <%=
        form_for(
            :attachment_metadata,
            url: { action: "create" },
            html: { multipart: true }
        ) do |form|
        %>

2. Use the file_field helper with :uploaded_data as the field name. Example:

       <%= form.file_field :uploaded_data %>

Displaying uploaded images is made easy by the +public_filename+ method of the
ActiveRecord attachment objects using file system, s3, and Cloud Files storage.

== public_filename(thumbnail = nil)

Returns the public path to the file.  If a thumbnail prefix is specified it will
return the public file path to the corresponding thumbnail.  Examples:

    attachment_obj.public_filename         #=> /attachments/2/file.jpg
    attachment_obj.public_filename(:thumb) #=> /attachments/2/file_thumb.jpg
    attachment_obj.public_filename(:small) #=> /attachments/2/file_small.jpg

When serving files from database storage, doing more than simply downloading the
file is beyond the scope of this document.

= attachment_fu controllers

There are two considerations to take into account when using attachment_fu in
controllers.

The first is when the files have no publicly accessible path and need to be
downloaded through an action.

Example:

    def readme
        send_file(
            '/path/to/readme.txt',
            type: 'plain/text',
            disposition: 'inline'
        )
    end

See the possible values for send_file for reference.

The second is when saving the file when submitted from a form.

Example in view:

    <%= form.file_field :attachable, :uploaded_data %>

Example in controller:

    def create
        @attachable_file = AttachmentMetadataModel.new(params[:attachable])
        if @attachable_file.save
            flash[:notice] = 'Attachment was successfully created.'
            redirect_to(attachable_url(@attachable_file))
        else
            redirect_to(action: 'new')
        end
    end

= attachement_fu scripting

You may wish to import a large number of images or attachments. The following
example shows how to upload a file from a script.

    #!/usr/bin/env ./script/runner

    # required to use ActionController::TestUploadedFile
    require 'action_controller'
    require 'action_controller/test_process.rb'

    path = "./public/images/x.jpg"

    # `mimetype` is a string like "image/jpeg". One way to get the mimetype for
    # a given file on a UNIX system: mimetype = `file -ib #{path}`.gsub(/\n/,"")
    mimetype = "image/jpeg"

    # This will "upload" the file at path and create the new model.
    @attachable = AttachmentMetadataModel.new(
        uploaded_data: ActionController::TestUploadedFile.new(path, mimetype)
    )
    @attachable.save


CHANGELOG
* Jul 15, 2015 *
* Pull in changes from brauliobo to fix protected_attributes for Rails 4.2 *
* Acknowlegement also to malmckay who provided another elegant fix to the protected_attributes problem and I want to thank him for it, but I had to pick 1 *

* Apr 27, 2015 *
* Pull in changes from benstein to gracefully handle nil filename

* Jan 27, 2015 *
* Pull in changes from bacrossland to fix test suite *

* Sep 23, 2014 *
* Pull in changes from corlissc to Refactor DbFileBackend for Rails 4 AREL conformity *

* May 13, 2014 *
* Pull in changes from bacrossland to fix thumbnails for S3 storage

* Apr 17, 2014 *
* Pull in changes from jasonbuehler to fix Rails 4 errors

* Jan 7, 2014 *
* Pull in changes from twelvelabs to fix gem use in Rails 2

* Aug 22, 2013 *
* Added cache-control header option and Ruby 1.9 fix for S3 from tricycle

* Aug 21, 2013 *
* Added S3 :encrypted_storage option support from cschulte22

* Jul 5, 2013
* Pull in changes from lchimonji10 to reformat README to RDoc format

* Apr 10, 2013 *
* Ruby 2 compatibility fix
* Removed some lingering occurrences of RAILS_ROOT and RAILS_ENV
* clean up of gemspec file
* no longer package unnecessary files like the test files

* Mar 12, 2013 *
* Pull in changes from https://github.com/aalong/attachment_fu to use Ruby 1.9.2 Tempfile naming strategy to fix conflicts with Sprockets (see https://github.com/aalong/attachment_fu/commit/938ec3b8597fbf82b1de8c98de12c4688463450a)

* Feb 19, 2013 *
* Pull in changes from https://github.com/itkin/attachment_fu fork to fix mini magick

* Nov 15, 2012 *
* Removed 'crop:' option as already included when prefixing geometry with a 'c'.  Add note to docs.
* Added option to sharpen resized images ':sharpen_on_resize'

* Oct 26, 2012 *
* Merged with https://github.com/tdd/attachment_fu fork to include GEM support for use in Rails 3.2, and bumped version to 3.2.x to indicate Rails 3.2 support.

* Oct 25, 2012 *
* Fix to support Ruby 1.9.3
* Image cropping support (append 'crop:' to desired file size when resizing)

* Mar 15, 2010 *
* Added a symbol syntax for parent-type-based size spec: calls the corresponding method on the current asset instance to
  get a Hash of required thumbnails.  Lets us dynamically specify what would otherwise be hard-coded, which is useful
  when the set of thumbnails for a given parent type varies depending on the usage context.

* Aug 6, 2009 *
* JPEG quality control finalized across processors (although CoreImage applies it in a rather fuzzy-logic way), with tests.

* Aug 4, 2009 *
* Supports the :aspect/'!' geometry flag in all processors
  (thanks to http://www.deepcalm.com/writing/cropped-thumbnails-in-attachment_fu-using-imagescience)
* JPEG quality control on thumbnailing/resizing (still buggy on Rmagick/MiniMagick though).
* Moves from GIF to PNG regardless of the source file extension's case (used to require lowercase)
* Auto-orients image (if EXIF suggests it) prior to processing with RMagickProcessor
* Fixes non-image upload tests (both regular files and Merb files)
* Fixes obsolete failures on RMagick tests (aspect_ratio tested but not initialized anymore due to new callback architecture

* Apr 17 2008 *
* amazon_s3.yml is now passed through ERB before being passed to AWS::S3 [François Beausoleil]

* Mar 22 2008 *
* Some tweaks to support Rails 2.0 and Rails 2.1 due to ActiveSupport::Callback changes.  
  Thanks to http://blog.methodmissing.com/2008/1/19/edge-callback-refactorings-attachment_fu/

* Feb. 26, 2008 *
* remove breakpoint from test_helper, makes test suite crazy (at least Rails 2+) [Rob Sanheim]
* make S3 test really optional [Rob Sanheim]

* Nov 27, 2007 *
* Handle properly ImageScience thumbnails resized from a gif file [Matt Aimonetti]
* Save thumbnails file size properly when using ImageScience [Matt Aimonetti]
* fixed s3 config file loading with latest versions of Rails [Matt Aimonetti]

* April 2, 2007 *

* don't copy the #full_filename to the default #temp_paths array if it doesn't exist
* add default ID partitioning for attachments
* add #binmode call to Tempfile (note: ruby should be doing this!) [Eric Beland]
* Check for current type of :thumbnails option.
* allow customization of the S3 configuration file path with the :s3_config_path option.
* Don't try to remove thumbnails if there aren't any.  Closes #3 [ben stiglitz]

* BC * (before changelog)

* add default #temp_paths entry [mattly]
* add MiniMagick support to attachment_fu [Isacc]
* update #destroy_file to clear out any empty directories too [carlivar]
* fix references to S3Backend module [Hunter Hillegas]
* make #current_data public with db_file and s3 backends [ebryn]
* oops, actually svn add the files for s3 backend. [Jeffrey Hardy]
* experimental s3 support, egad, no tests.... [Jeffrey Hardy]
* doh, fix a few bad references to ActsAsAttachment [sixty4bit]


lib/pothoven-attachment_fu.rb

if defined?(Rails::Engine)
  # Rails >= 3
  class Engine < Rails::Engine
    # Mimic old vendored plugin behavior, attachment_fu/lib is autoloaded.
    config.autoload_paths << File.expand_path("..", __FILE__)

    initializer "attachment_fu" do
      require 'geometry'

      ActiveRecord::Base.send(:extend, Technoweenie::AttachmentFu::ActMethods)
      Technoweenie::AttachmentFu.tempfile_path = ATTACHMENT_FU_TEMPFILE_PATH if Object.const_defined?(:ATTACHMENT_FU_TEMPFILE_PATH)
      FileUtils.mkdir_p Technoweenie::AttachmentFu.tempfile_path
    end
  end
else
  # Rails <= 2
  require 'geometry'

  ActiveRecord::Base.send(:extend, Technoweenie::AttachmentFu::ActMethods)
  Technoweenie::AttachmentFu.tempfile_path = ATTACHMENT_FU_TEMPFILE_PATH if Object.const_defined?(:ATTACHMENT_FU_TEMPFILE_PATH)
  FileUtils.mkdir_p Technoweenie::AttachmentFu.tempfile_path

  $:.unshift(File.expand_path("../vendor", __FILE__))
end


lib/technoweenie/attachment_fu/backends/s3_backend.rb
module Technoweenie # :nodoc:
  module AttachmentFu # :nodoc:
    module Backends
      # = AWS::S3 Storage Backend
      #
      # Enables use of {Amazon's Simple Storage Service}[http://aws.amazon.com/s3] as a storage mechanism
      #
      # == Requirements
      #
      # Requires the {AWS::S3 Library}[http://amazon.rubyforge.org] for S3 by Marcel Molina Jr. installed either
      # as a gem or a as a Rails plugin.
      #
      # == Configuration
      #
      # Configuration is done via <tt>#{Rails.root}/config/amazon_s3.yml</tt> and is loaded according to the <tt>#{Rails.env}</tt>.
      # The minimum connection options that you must specify are a bucket name, your access key id and your secret access key.
      # If you don't already have your access keys, all you need to sign up for the S3 service is an account at Amazon.
      # You can sign up for S3 and get access keys by visiting http://aws.amazon.com/s3.
      #
      # If you wish to use Amazon CloudFront to serve the files, you can also specify a distibution domain for the bucket.
      # To read more about CloudFront, visit http://aws.amazon.com/cloudfront
      #
      # Example configuration (#{Rails.root}/config/amazon_s3.yml)
      #
      #   development:
      #     bucket_name: appname_development
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #     distribution_domain: XXXX.cloudfront.net
      #
      #   test:
      #     bucket_name: appname_test
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #     distribution_domain: XXXX.cloudfront.net
      #
      #   production:
      #     bucket_name: appname
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #     distribution_domain: XXXX.cloudfront.net
      #
      # You can change the location of the config path by passing a full path to the :s3_config_path option.
      #
      #   has_attachment :storage => :s3, :s3_config_path => (#{Rails.root} + '/config/s3.yml')
      #
      # === Required configuration parameters
      #
      # * <tt>:access_key_id</tt> - The access key id for your S3 account. Provided by Amazon.
      # * <tt>:secret_access_key</tt> - The secret access key for your S3 account. Provided by Amazon.
      # * <tt>:bucket_name</tt> - A unique bucket name (think of the bucket_name as being like a database name).
      #
      # If any of these required arguments is missing, a MissingAccessKey exception will be raised from AWS::S3.
      #
      # == About bucket names
      #
      # Bucket names have to be globaly unique across the S3 system. And you can only have up to 100 of them,
      # so it's a good idea to think of a bucket as being like a database, hence the correspondance in this
      # implementation to the development, test, and production environments.
      #
      # The number of objects you can store in a bucket is, for all intents and purposes, unlimited.
      #
      # === Optional configuration parameters
      #
      # * <tt>:server</tt> - The server to make requests to. Defaults to <tt>s3.amazonaws.com</tt>.
      # * <tt>:port</tt> - The port to the requests should be made on. Defaults to 80 or 443 if <tt>:use_ssl</tt> is set.
      # * <tt>:use_ssl</tt> - If set to true, <tt>:port</tt> will be implicitly set to 443, unless specified otherwise. Defaults to false.
      # * <tt>:distribution_domain</tt> - The CloudFront distribution domain for the bucket.  This can either be the assigned
      #     distribution domain (ie. XXX.cloudfront.net) or a chosen domain using a CNAME. See CloudFront for more details.
      #
      # == Usage
      #
      # To specify S3 as the storage mechanism for a model, set the acts_as_attachment <tt>:storage</tt> option to <tt>:s3</tt>.
      #
      #   class Photo < ActiveRecord::Base
      #     has_attachment :storage => :s3
      #   end
      #
      # === Customizing the path
      #
      # By default, files are prefixed using a pseudo hierarchy in the form of <tt>:table_name/:id</tt>, which results
      # in S3 urls that look like: http(s)://:server/:bucket_name/:table_name/:id/:filename with :table_name
      # representing the customizable portion of the path. You can customize this prefix using the <tt>:path_prefix</tt>
      # option:
      #
      #   class Photo < ActiveRecord::Base
      #     has_attachment :storage => :s3, :path_prefix => 'my/custom/path'
      #   end
      #
      # Which would result in URLs like <tt>http(s)://:server/:bucket_name/my/custom/path/:id/:filename.</tt>
      #
      # === Using different bucket names on different models
      #
      # By default the bucket name that the file will be stored to is the one specified by the
      # <tt>:bucket_name</tt> key in the amazon_s3.yml file.  You can use the <tt>:bucket_key</tt> option
      # to overide this behavior on a per model basis.  For instance if you want a bucket that will hold
      # only Photos you can do this:
      #
      #   class Photo < ActiveRecord::Base
      #     has_attachment :storage => :s3, :bucket_key => :photo_bucket_name
      #   end
      #
      # And then your amazon_s3.yml file needs to look like this.
      #
      #   development:
      #     bucket_name: appname_development
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #
      #   test:
      #     bucket_name: appname_test
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #
      #   production:
      #     bucket_name: appname
      #     photo_bucket_name: appname_photos
      #     access_key_id: <your key>
      #     secret_access_key: <your key>
      #
      #  If the bucket_key you specify is not there in a certain environment then attachment_fu will
      #  default to the <tt>bucket_name</tt> key.  This way you only have to create special buckets
      #  this can be helpful if you only need special buckets in certain environments.
      #
      # === Permissions
      #
      # By default, files are stored on S3 with public access permissions. You can customize this using
      # the <tt>:s3_access</tt> option to <tt>has_attachment</tt>. Available values are
      # <tt>:private</tt>, <tt>:public_read_write</tt>, and <tt>:authenticated_read</tt>.
      #
      # === Other options
      #
      # Of course, all the usual configuration options apply, such as content_type and thumbnails:
      #
      #   class Photo < ActiveRecord::Base
      #     has_attachment :storage => :s3, :content_type => ['application/pdf', :image], :resize_to => 'x50'
      #     has_attachment :storage => :s3, :thumbnails => { :thumb => [50, 50], :geometry => 'x50' }
      #   end
      #
      # === Accessing S3 URLs
      #
      # You can get an object's URL using the s3_url accessor. For example, assuming that for your postcard app
      # you had a bucket name like 'postcard_world_development', and an attachment model called Photo:
      #
      #   @postcard.s3_url # => http(s)://s3.amazonaws.com/postcard_world_development/photos/1/mexico.jpg
      #
      # The resulting url is in the form: http(s)://:server/:bucket_name/:table_name/:id/:file.
      # The optional thumbnail argument will output the thumbnail's filename (if any).
      #
      # Additionally, you can get an object's base path relative to the bucket root using
      # <tt>base_path</tt>:
      #
      #   @photo.file_base_path # => photos/1
      #
      # And the full path (including the filename) using <tt>full_filename</tt>:
      #
      #   @photo.full_filename # => photos/
      #
      # Niether <tt>base_path</tt> or <tt>full_filename</tt> include the bucket name as part of the path.
      # You can retrieve the bucket name using the <tt>bucket_name</tt> method.
      #
      # === Accessing CloudFront URLs
      #
      # You can get an object's CloudFront URL using the cloudfront_url accessor.  Using the example from above:
      # @postcard.cloudfront_url # => http://XXXX.cloudfront.net/photos/1/mexico.jpg
      #
      # The resulting url is in the form: http://:distribution_domain/:table_name/:id/:file
      #
      # If you set :cloudfront to true in your model, the public_filename will be the CloudFront
      # URL, not the S3 URL.
      module S3Backend
        class RequiredLibraryNotFoundError < StandardError; end
        class ConfigFileNotFoundError < StandardError; end

        def self.included(base) #:nodoc:
          mattr_reader :bucket_name, :s3_config

          begin
            require 'aws/s3'
            include AWS::S3
          rescue LoadError
            raise RequiredLibraryNotFoundError.new('AWS::S3 could not be loaded')
          end

          begin
            @@s3_config_path = base.attachment_options[:s3_config_path] || File.join(Rails.root, 'config', 'amazon_s3.yml')
            @@s3_config = @@s3_config = YAML.load(ERB.new(File.read(@@s3_config_path)).result)[Rails.env].symbolize_keys
          #rescue
          #  raise ConfigFileNotFoundError.new('File %s not found' % @@s3_config_path)
          end

          bucket_key = base.attachment_options[:bucket_key]

          if bucket_key and s3_config[bucket_key.to_sym]
            eval_string = "def bucket_name()\n  \"#{s3_config[bucket_key.to_sym]}\"\nend"
          else
            eval_string = "def bucket_name()\n  \"#{s3_config[:bucket_name]}\"\nend"
          end
          base.class_eval(eval_string, __FILE__, __LINE__)

          Base.establish_connection!(s3_config.slice(:access_key_id, :secret_access_key, :server, :port, :use_ssl, :persistent, :proxy))

          # Bucket.create(@@bucket_name)

          base.before_update :rename_file
        end

        def self.protocol
          @protocol ||= s3_config[:use_ssl] ? 'https://' : 'http://'
        end

        def self.hostname
          @hostname ||= s3_config[:server] || AWS::S3::DEFAULT_HOST
        end

        def self.port_string
          @port_string ||= (s3_config[:port].nil? || s3_config[:port] == (s3_config[:use_ssl] ? 443 : 80)) ? '' : ":#{s3_config[:port]}"
        end

        def self.distribution_domain
          @distribution_domain = s3_config[:distribution_domain]
        end

        module ClassMethods
          def s3_protocol
            Technoweenie::AttachmentFu::Backends::S3Backend.protocol
          end

          def s3_hostname
            Technoweenie::AttachmentFu::Backends::S3Backend.hostname
          end

          def s3_port_string
            Technoweenie::AttachmentFu::Backends::S3Backend.port_string
          end

          def cloudfront_distribution_domain
            Technoweenie::AttachmentFu::Backends::S3Backend.distribution_domain
          end
        end

        # Overwrites the base filename writer in order to store the old filename
        def filename=(value)
          @old_filename = filename unless filename.nil? || @old_filename
          write_attribute :filename, sanitize_filename(value)
        end

        # The attachment ID used in the full path of a file
        def attachment_path_id
          ((respond_to?(:parent_id) && parent_id) || id).to_s
        end

        # The pseudo hierarchy containing the file relative to the bucket name
        # Example: <tt>:table_name/:id</tt>
        def base_path
          File.join(attachment_options[:path_prefix], attachment_path_id)
        end

        # The full path to the file relative to the bucket name
        # Example: <tt>:table_name/:id/:filename</tt>
        def full_filename(thumbnail = nil)
          File.join(base_path, thumbnail_name_for(thumbnail))
        end

        # All public objects are accessible via a GET request to the S3 servers. You can generate a
        # url for an object using the s3_url method.
        #
        #   @photo.s3_url
        #
        # The resulting url is in the form: <tt>http(s)://:server/:bucket_name/:table_name/:id/:file</tt> where
        # the <tt>:server</tt> variable defaults to <tt>AWS::S3 URL::DEFAULT_HOST</tt> (s3.amazonaws.com) and can be
        # set using the configuration parameters in <tt>#{Rails.root}/config/amazon_s3.yml</tt>.
        #
        # The optional thumbnail argument will output the thumbnail's filename (if any).
        def s3_url(thumbnail = nil)
          File.join(s3_protocol + s3_hostname + s3_port_string, bucket_name, full_filename(thumbnail))
        end

        # All public objects are accessible via a GET request to CloudFront. You can generate a
        # url for an object using the cloudfront_url method.
        #
        #   @photo.cloudfront_url
        #
        # The resulting url is in the form: <tt>http://:distribution_domain/:table_name/:id/:file</tt> using
        # the <tt>:distribution_domain</tt> variable set in the configuration parameters in <tt>#{Rails.root}/config/amazon_s3.yml</tt>.
        #
        # The optional thumbnail argument will output the thumbnail's filename (if any).
        def cloudfront_url(thumbnail = nil)
          s3_protocol + cloudfront_distribution_domain + "/" + full_filename(thumbnail)
        end

        def public_filename(*args)
          if attachment_options[:cloudfront]
            cloudfront_url(args)
          else
            s3_url(args)
          end
        end

        # All private objects are accessible via an authenticated GET request to the S3 servers. You can generate an
        # authenticated url for an object like this:
        #
        #   @photo.authenticated_s3_url
        #
        # By default authenticated urls expire 5 minutes after they were generated.
        #
        # Expiration options can be specified either with an absolute time using the <tt>:expires</tt> option,
        # or with a number of seconds relative to now with the <tt>:expires_in</tt> option:
        #
        #   # Absolute expiration date (October 13th, 2025)
        #   @photo.authenticated_s3_url(:expires => Time.mktime(2025,10,13).to_i)
        #
        #   # Expiration in five hours from now
        #   @photo.authenticated_s3_url(:expires_in => 5.hours)
        #
        # You can specify whether the url should go over SSL with the <tt>:use_ssl</tt> option.
        # By default, the ssl settings for the current connection will be used:
        #
        #   @photo.authenticated_s3_url(:use_ssl => true)
        #
        # Finally, the optional thumbnail argument will output the thumbnail's filename (if any):
        #
        #   @photo.authenticated_s3_url('thumbnail', :expires_in => 5.hours, :use_ssl => true)
        def authenticated_s3_url(*args)
          options   = args.extract_options!
          options[:expires_in] = options[:expires_in].to_i if options[:expires_in]
          thumbnail = args.shift
          S3Object.url_for(full_filename(thumbnail), bucket_name, options)
        end

        def create_temp_file
          write_to_temp_file current_data
        end

        def current_data
          if attachment_options[:encrypted_storage] && self.respond_to?(:encryption_key) && self.encryption_key != nil
            EncryptedData.decrypt_data(S3Object.value(full_filename, bucket_name), self.encryption_key)
          else
            S3Object.value full_filename, bucket_name
          end
        end

        def s3_protocol
          Technoweenie::AttachmentFu::Backends::S3Backend.protocol
        end

        def s3_hostname
          Technoweenie::AttachmentFu::Backends::S3Backend.hostname
        end

        def s3_port_string
          Technoweenie::AttachmentFu::Backends::S3Backend.port_string
        end

        def cloudfront_distribution_domain
          Technoweenie::AttachmentFu::Backends::S3Backend.distribution_domain
        end

        protected
          # Called in the after_destroy callback
          def destroy_file
            S3Object.delete full_filename, bucket_name
          end

          def rename_file
            return unless @old_filename && @old_filename != filename

            old_full_filename = File.join(base_path, @old_filename)

            S3Object.rename(
              old_full_filename,
              full_filename,
              bucket_name,
              :access => attachment_options[:s3_access]
            )

            @old_filename = nil
            true
          end

          def save_to_storage
            if save_attachment?
              if attachment_options[:encrypted_storage]
                S3Object.store(
                               full_filename,
                               (temp_path ? File.open(temp_path) : temp_data),
                               bucket_name,
                               :content_type => content_type,
                               :access => attachment_options[:s3_access],
                               'x-amz-server-side-encryption' => 'AES256',
                               'Content-Disposition' => "attachment; filename=\"#{filename}\""
                               )
              else
                S3Object.store(
                               full_filename,
                               (temp_path ? File.open(temp_path) : temp_data),
                               bucket_name,
                               :content_type => content_type,
                               :access => attachment_options[:s3_access]
                               )
              end
            end

            @old_filename = nil
            true
          end
      end
    end
  end
end
